lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -598,6 +600,61 @@ newDBLayer trace defaultFieldValues mDatabaseFile timeInterpreter = do
             migrateAll
             trace
             mDatabaseFile
+
+    -- NOTE1
+    -- We cache the latest checkpoint for read operation such that we prevent
+    -- needless marshalling and unmarshalling with the database. Many handlers
+    -- dealing with the database are actually in the form of:
+    --
+    --   - read latest CP
+    --   - write latest CP
+    --
+    -- When chaining them, we end up paying the cost of unmarshalling data from
+    -- the database one extra time. That doesn't matter much for small wallets
+    -- because the time needed to unmarshall data is relatively negligible. For
+    -- large wallets however, this has a massive performance impact.
+    --
+    -- Instead, the cache now retains the Haskell data-types in-memory to
+    -- short-circuit the most frequent database lookups.
+    --
+    -- NOTE2
+    -- We use an IORef here without fearing race-conditions because every
+    -- database query can only be run within calls to `atomically` which
+    -- enforces that there's only a single thread executing a given
+    -- `SqlPersistT`.
+    cache <- newIORef Map.empty
+
+    let readCache :: W.WalletId -> SqlPersistT IO (Maybe (W.Wallet s))
+        readCache wid = Map.lookup wid <$> liftIO (readIORef cache)
+
+    let clearCache :: W.WalletId -> SqlPersistT IO ()
+        clearCache wid = liftIO $ modifyIORef cache $ Map.delete wid
+
+    let writeCache :: W.WalletId -> W.Wallet s -> SqlPersistT IO ()
+        writeCache wid cp = liftIO $ modifyIORef cache $ Map.alter alter wid
```

Comment:
Is there a chance of buildup of thunks here? Though I suppose we'd notice that when testing this

Reply:
Thought of it and:

- Map is strict, so values are evaluated before they are put in the Map 
- We never just write in the cache without reading and evaluating data from it first. 

So no, I don't think there's any chance that large thunks will be built here. I could switch to `modifyIORef'` for the tranquility of mind, we don't need the lazyness anyway since we have in principle already evaluated the cache to WHNF (and that's actually part of why we want to cache it!)